sdk/src/Core/Amazon.Util/AWSSDKUtils.cs

```diff
@@ -199,8 +199,22 @@ private static string DetermineValidPathCharacters()
         /// <returns></returns>
         public static string GetExtension(string path)
         {
-            if (path == null)
+            if (path is null)
```

Comment:
Why not simply replace this with https://learn.microsoft.com/en-us/dotnet/api/system.io.path.getextension?view=net-8.0?

It seems it covers all the edge cases already and also has the directory and alt directory separator char. That version is already heavily optimized, and then the complexity would go away of having to manage TFM specific code.

Reply:
@danielmarbach  I actually gave that a try :)

Benchmarking yielded numbers similar to the existing AWSSDK code, which as it turns out use [very similar algorithms](https://github.com/dotnet/runtime/blob/5535e31a712343a63f5d7d796cd874e563e5ac14/src/libraries/System.Private.CoreLib/src/System/IO/Path.cs#L197C13-L212C45). It has indeed been heavily optimized for the filesystem "happy path" where a short file extension exists.

There is one notable difference though in that the current dotnet runtime is _more permissive_ than the current AWSSDK implementation. The AWSSDK will currently return an empty string if any of `/`, `\` or `:` are present, regardless of underlying runtime. The current built-in Windows dotnet runtime, however, will [no longer detect volume separator](https://github.com/dotnet/runtime/blob/main/src/libraries/Common/src/System/IO/PathInternal.Windows.cs#L280-L283) while the Unix runtime will [not detect a volume separator or a backslash](https://github.com/dotnet/runtime/blob/main/src/libraries/Common/src/System/IO/PathInternal.Unix.cs#L27-L33) - only forward slash. I believe the the opposite may've been true for the full framework side, where the built-in one was _too restrictive_ in that it may throw for various invalid filesystem characters like `*`, `|` etc.

@normj Reaching out as I see you've been requested for review here. This does introduce added complexity and I recognize that optimizing for GUID (or similar) keys may be a less-common use case than I may imagine. I'm happy to go in any direction you like here. Note that if optimizing for GUIDs were of interest then this could be taken a bit further by adding this after the first `LastIndexOf()` call effectively bring the GUID case down to around 2ns on my machine at the expense of another 0.5ns for the cases where a file extension _does_ exist.

```diff
int extensionIndex = path.AsSpan().LastIndexOf('.');
+if (extensionIndex == -1) {
+  return string.Empty;
+}

int directoryIndex = path.AsSpan().LastIndexOfAny('/', '\\', ':');
```

Reply:
Are you saying using the extension method that @danielmarbach shows no significant performance boost over what is shipped in the SDK today?

The code change you have for .NET 8+ is pretty simple and straight forward so I'm good with taking the change assuming we are saying this approach is still faster then the extension method coming from the framework. I would ask that you add a comment in the `#if` explaining how newer versions of .NET have an optimized version of `LastIndexOf`. Basically give a hint for feature reads of the code wondering why we have 2 separate implementations.

Reply:
Apologies, my last comment was a little verbose as I'd been rushing to send it before the bottom of the hour here.

> Are you saying using the extension method that @danielmarbach shows no significant performance boost over what is shipped in the SDK today?

@normj 
This is correct. I've re-benchmarked to include the built-in .NET version as well as including the "early return" in this version. I've updated the original PR description to include the benchmarking code and latest results. I'll push the changes shortly, after resolving some conflicts.